File: discovery_engine_core/ai/xayn-ai/src/ranker/document.rs
// Copyright 2022 Xayn AG
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, version 3.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use chrono::NaiveDateTime;

use crate::{embedding::utils::Embedding, DocumentId};

pub trait Document {
    /// Gets the document id.
    fn id(&self) -> DocumentId;

    /// Gets the SMBert embedding of the document.
    fn smbert_embedding(&self) -> &Embedding;

    /// Gets the publishing date.
    fn date_published(&self) -> NaiveDateTime;
}

#[cfg(test)]
pub(super) struct TestDocument {
    pub(super) id: DocumentId,
    pub(super) smbert_embedding: Embedding,
    pub(super) date_published: NaiveDateTime,
}

#[cfg(test)]
impl TestDocument {
    pub(super) fn new(id: u128, embedding: impl Into<Embedding>, date_published: &str) -> Self {
        Self {
            id: DocumentId::from_u128(id),
            smbert_embedding: embedding.into(),
            date_published: NaiveDateTime::parse_from_str(date_published, "%Y-%m-%d %H:%M:%S")
                .unwrap(),
        }
    }
}

#[cfg(test)]
impl Document for TestDocument {
    fn id(&self) -> DocumentId {
        self.id
    }

    fn smbert_embedding(&self) -> &Embedding {
        &self.smbert_embedding
    }

    fn date_published(&self) -> NaiveDateTime {
        self.date_published
    }
}

File: discovery_engine/lib/src/ffi/types/migration/document.dart
// Copyright 2022 Xayn AG
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, version 3.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import 'dart:ffi';

import 'package:xayn_discovery_engine/src/domain/models/active_data.dart';
import 'package:xayn_discovery_engine/src/domain/models/document.dart';
import 'package:xayn_discovery_engine/src/domain/models/view_mode.dart';
import 'package:xayn_discovery_engine/src/ffi/genesis.ffigen.dart';
import 'package:xayn_discovery_engine/src/ffi/load_lib.dart' show ffi;
import 'package:xayn_discovery_engine/src/ffi/types/date_time.dart';
import 'package:xayn_discovery_engine/src/ffi/types/document/news_resource.dart';
import 'package:xayn_discovery_engine/src/ffi/types/document/user_reaction.dart';
import 'package:xayn_discovery_engine/src/ffi/types/duration.dart';
import 'package:xayn_discovery_engine/src/ffi/types/embedding.dart';
import 'package:xayn_discovery_engine/src/ffi/types/primitives.dart';
import 'package:xayn_discovery_engine/src/ffi/types/uuid.dart';

class MigrationDocument {
  final Document document;
  final ActiveDocumentData? activeData;

  MigrationDocument(this.document, this.activeData);

  void writeNative(Pointer<RustMigrationDocument> place) {
    document.documentId.writeNative(ffi.migration_document_place_of_id(place));
    document.stackId
        .writeNative(ffi.migration_document_place_of_stack_id(place));
    // ignore: deprecated_member_use_from_same_package
    (activeData?.smbertEmbedding)
        .writeNative(ffi.migration_document_place_of_smbert_embedding(place));
    document.userReaction
        .writeNative(ffi.migration_document_place_of_reaction(place));
    document.resource
        .writeNative(ffi.migration_document_place_of_resource(place));
    ffi.init_migration_document_is_active_at(place, document.isActive ? 1 : 0);
    ffi.init_migration_document_is_searched_at(
      place,
      document.isSearched ? 1 : 0,
    );

    // ignore: deprecated_member_use_from_same_package
    document.batchIndex
        .writeNative(ffi.migration_document_place_of_batch_index(place));

    // ignore: deprecated_member_use_from_same_package
    document.timestamp
        .writeNative(ffi.migration_document_place_of_timestamp(place));

    (activeData?.viewTime[DocumentViewMode.web])
        .writeNative(ffi.migration_document_place_of_web_view_time(place));
    (activeData?.viewTime[DocumentViewMode.reader])
        .writeNative(ffi.migration_document_place_of_reader_view_time(place));
    (activeData?.viewTime[DocumentViewMode.story])
        .writeNative(ffi.migration_document_place_of_story_view_time(place));
  }
}
